### PR TITLE
fix(llm): serialize OpenAI tool call messages

### PR DESCRIPTION
**Pull request by Iams4kura**
## What Changed

- Add a typed provider-layer OpenAI message adapter that emits assistant tool
  calls with `type: "function"` and JSON-string `function.arguments`.
- Use that serializer for OpenAI, Astraflow (including AstraflowCN), and Atlas
  requests while keeping the provider-neutral, Claude, and Ollama formats unchanged.
- Reject non-standard or non-serializable function arguments before they reach the
  client, with a stable provider-boundary error.
- Reject tool-result messages without a `tool_call_id` before making a request.
- Add a provider-boundary regression test covering the exact payload, strict JSON
  failures, and no-client-call behavior for all three implementations.

## Why This Change

After a model requested a tool, the ReAct loop replayed the assistant tool-call
history through `Message.to_dict()`. That representation omitted the required
tool-call `type` and kept `function.arguments` as an object instead of the JSON
string expected by the OpenAI Chat Completions wire format. A follow-up request
could therefore be rejected before the tool loop completed.

The generic serializer is intentionally unchanged because other providers use a
different message shape. An older MiniMax-specific implementation in #2510 uses
the same wire-format concepts, but it does not cover these existing providers or
the shared message boundary.

## Testing Done

- [ ] Manual testing completed
- [x] Automated tests pass locally (`node tests/run-all.js`)
- [x] Edge cases considered and tested

- Regression before the production change: `python -m pytest tests/test_provider_tools.py -q` — 3 expected failures, one for each provider.
- Regression after the fix: the same command — 8 passed.
- Post-review regression — 8 passed, including NaN and non-serializable arguments.
- Affected Python suite — 37 passed.
- Python project suite (`tests/test_*.py`, excluding integration) — 88 passed.
- Ruff and mypy — passed.
- Node 20.19.0 `npm test` — 4,159 passed, 0 failed.
- Node 20.19.0 `npm run coverage` — 4,159 passed; 89.04% statements/lines, 94.39% functions, 80.87% branches.
- Node 20.19.0 `npm run lint` — passed.

## Type of Change
- [x] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [ ] `docs:` Documentation
- [ ] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [x] Shell scripts pass shellcheck (if applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

No shell files changed, and the repository does not configure a pre-commit hook
for this Python path.

## If you changed dependencies or `package.json` (`bin` / `files` / deps)
- [ ] Ran `yarn install --mode=update-lockfile` and committed the `yarn.lock` change. CI runs Yarn in hardened mode on public PRs and fails if the lockfile would be modified, so an out of date `yarn.lock` breaks the build even when nothing else is wrong.

Not applicable: no dependency, package manifest, or lockfile changed.

## If you added a skill, command, agent, hook, or CLI tool
- [ ] Registered in `package.json` (`bin` and `files`), `manifests/install-components.json`, `manifests/install-modules.json`, and `agent.yaml`
- [ ] Regenerated the catalog (`npm run catalog:sync`) and command registry (`npm run command-registry:write`)
- [ ] Updated the docs tables it belongs in (`README.md`, `COMMANDS-QUICK-REF.md`, `docs/COMMAND-AGENT-MAP.md`)
- [ ] If it ships a new script path, added it to the publish surface allowlist (`tests/scripts/npm-publish-surface.test.js`)
- [ ] Cross-harness surfaces updated if applicable (for Codex, `.agents/skills/<name>/` plus `agents/openai.yaml`; the Codex frontmatter validator allows only `name`, `description`, `metadata`, `license`, `allowed-tools`, so drop keys like `version` from that copy)
- [ ] Full gauntlet passes locally (`npm test`)

Not applicable: this change adds no skill, command, agent, hook, or CLI surface.

## Documentation
- [ ] Updated relevant documentation
- [ ] Added comments for complex logic
- [ ] README updated (if needed)

Not required: this is an internal wire-format correction covered by a regression
test and does not change the documented user interface.

## Scope

5 files changed, +184 / -4 lines. No generated, dependency, or lock files changed.

### CODE REVIEW

File: src/llm/core/types.py [hunk 1]
```diff
@@ -44,6 +45,22 @@ def to_dict(self) -> dict[str, Any]:
...
+                    "id": tc.id,
+                    "type": "function",
+                    "function": {
+                        "name": tc.name,
+                        "arguments": json.dumps(tc.arguments),
+                    },
+                }
+                for tc in self.tool_calls
+            ]
+        return result
```
[the remark this answers is not in the document]
Comment (Iams4kura) [reply src/llm/core/types.py]: Addressed in 4a9bdff. The mapping now lives in `llm.providers.openai_compat`, constructs role-specific `ChatCompletion*Param` payloads using public types available since OpenAI 1.30.0, and is shared by OpenAI, Astraflow, and Atlas. `core/types.py` now has no net diff. Verified with the provider regression (8 passed), affected tests (42), the Python suite (88), Ruff, mypy including this test, and the Node 20 suite/coverage (4,159 passed).

File: tests/test_provider_tools.py [hunk 2]
```diff
@@ -84,6 +86,55 @@ def test_openai_provider_serializes_tools_for_chat_completions():
     ]
 
 
+@pytest.mark.parametrize(
+    "provider_type",
+    [OpenAIProvider, AstraflowProvider, AtlasProvider],
+)
+def test_openai_compatible_provider_serializes_assistant_tool_calls(provider_type):
```
[the remark this answers is not in the document]
Comment (Iams4kura) [reply tests/test_provider_tools.py]: Addressed in 4a9bdff: the regression now has `@pytest.mark.unit`, a fully annotated signature, and passes mypy. I used `Callable[..., LLMProvider]` rather than `type[LLMProvider]` because the abstract base does not declare the concrete `api_key` constructor (mypy rejects that keyword through `type[LLMProvider]`); the fake-client `Any` cast is kept local to the test.

File: src/llm/providers/openai_compat.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+def _to_openai_tool_call_param(
+    tool_call: ToolCall,
+) -> ChatCompletionMessageToolCallParam:
+    """Convert a domain tool call to the OpenAI request schema."""
+    return {
+        "id": tool_call.id,
+        "type": "function",
+        "function": {
+            "name": tool_call.name,
+            "arguments": json.dumps(tool_call.arguments),
```
[the remark this answers is not in the document]
Comment (Iams4kura) [reply src/llm/providers/openai_compat.py]: Addressed in 8eedb3d: the shared adapter now uses json.dumps(..., allow_nan=False) and normalizes TypeError/ValueError to a stable provider-boundary ValueError while retaining the exception cause. The parameterized test covers both NaN and a non-serializable object across OpenAI, Astraflow, and Atlas, and confirms neither reaches the client. Final checks: provider tests 8 passed, affected tests 37 passed, Python project tests 88 passed, Ruff and mypy passed.

File: tests/test_provider_tools.py [hunk 4]
```diff
@@ -84,6 +89,70 @@ def test_openai_provider_serializes_tools_for_chat_completions():
...
+                        "name": "search",
+                        "arguments": '{"query": "ecc"}',
+                    },
+                },
+            ],
+        },
+        {"role": "tool", "content": "result", "tool_call_id": "call_1"},
+    ]
+
+    client.completions.params = None
```
[the remark this answers is not in the document]
Comment (Iams4kura) [reply tests/test_provider_tools.py]: Addressed in 8eedb3d: the test now preserves submitted_params and asserts object identity after every invalid request instead of mutating the fake client state. Because the fake create method assigns a fresh kwargs mapping on each call, unchanged identity proves the validation failed before a second client call. The updated provider test passes for OpenAI, Astraflow, and Atlas.

**Comment by chenhz01** [comment]
The adapter approach is right — one edge worth pinning down in `to_openai_message_param`: assistant messages replayed from history may already carry `function.arguments` as a JSON string (stored transcripts, or another provider's output fed back in). If the serializer unconditionally stringifies, an already-string value gets double-encoded into `'"..."'` and the API rejects the whole request. Guard it:

```py
if isinstance(args, str): json.loads(args)  # validate, pass through
else: args = json.dumps(args)
```

The provider-boundary test suite is the natural place for an idempotency case: serialize → re-serialize → assert identical payload. I also checked `src/llm/providers/ollama.py` on main — it posts to the native `/api/chat` endpoint, not an OpenAI-compatible one, so leaving it unchanged here is correct. Happy to take a crack at the idempotency test if useful.
